### PR TITLE
Add Intel Extension SKLearn flavor to Azure instances

### PR DESCRIPTION
**Pull request by sramakintel**
Currently, the README.md document provides an introduction to the details in the PR and how build and test the PR.
The Dockerfile parsing of `{{latest-image-tag}}` is done using `sed` command.

### CODE REVIEW

File: training/general/src/environments/sklearn-1.1/context/intel-extension/conda_dependencies.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+channels:
+- conda-forge
+- anaconda
+dependencies:
+- python=3.9
+- pip=21.3.1
+- pandas~=1.3.0
+- scipy~=1.7.0
+- numpy~=1.21.0
+- scikit-learn-intelex~={{latest-sklearnex-version}}
```
Comment (elesel) [inline training/general/src/environments/sklearn-1.1/context/intel-extension/conda_dependencies.yaml:11]: @savitamittal1 what do you think about adding this package to sklearn-1.1? If you're in favor then I'll have some other comments on the PR. We could also consider creating something like an azureml-contrib registry at some point, which could include this.

**Comment by rtanase** [comment]
@sramakintel, you shouldn't need to replace the macros in the dockerfile with sed, pls take a look [here] (https://github.com/Azure/azureml-assets/wiki/Environments#testing-environment-image-builds) and we can leave the macros in the dockerfile to automatically pick up latest versions (if that's the desired intent).

**Comment by sramakintel** [comment]
> @sramakintel, you shouldn't need to replace the macros in the dockerfile with sed, pls take a look [here] (https://github.com/Azure/azureml-assets/wiki/Environments#testing-environment-image-builds) and we can leave the macros in the dockerfile to automatically pick up latest versions (if that's the desired intent).

@rtanase : Thanks for the pointer. How to customize the tags? I see the link https://github.com/Azure/azureml-assets/wiki/Environments#template-tags mentions the tags that are supported. I have a new tag `{{latest-sklearnex-version}}` in here https://github.com/Azure/azureml-assets/blob/9e66d0cb92194e7353890ca13d092e780de07159/training/general/src/environments/sklearn-1.1/context/intel-extension/conda_dependencies.yaml#L11

File: training/general/src/environments/sklearn-1.1/context/intel-extension/conda_dependencies.yaml (hunk 1 above)
Comment (elesel) [inline training/general/src/environments/sklearn-1.1/context/intel-extension/conda_dependencies.yaml:11]: This will pull the latest version from PyPI when this environment is built:
```suggestion
- scikit-learn-intelex~={{latest-pypi-version}}
```

**Comment by elesel** [comment]
> > @sramakintel, you shouldn't need to replace the macros in the dockerfile with sed, pls take a look [here] (https://github.com/Azure/azureml-assets/wiki/Environments#testing-environment-image-builds) and we can leave the macros in the dockerfile to automatically pick up latest versions (if that's the desired intent).
> 
> @rtanase : Thanks for the pointer. How to customize the tags? I see the link https://github.com/Azure/azureml-assets/wiki/Environments#template-tags mentions the tags that are supported. I have a new tag `{{latest-sklearnex-version}}` in here
> 
> https://github.com/Azure/azureml-assets/blob/9e66d0cb92194e7353890ca13d092e780de07159/training/general/src/environments/sklearn-1.1/context/intel-extension/conda_dependencies.yaml#L11

Is the only substantive change in this PR the addition of the scikit-learn-intelex package? If so, @rtanase, should we just add the package to the existing sklearn-1.1 environment?

**Review by elesel — CHANGES_REQUESTED** [review]
This PR will need some updates after the decision is made on how to add the scikit-learn-intelex package to it.

File: training/general/src/environments/sklearn-1.1/context/intel-extension/conda_dependencies.yaml (hunk 1 above)
Comment (sramakintel) [reply training/general/src/environments/sklearn-1.1/context/intel-extension/conda_dependencies.yaml]: I see. Thanks. Removed the `sed` instructions from README.md.

**Comment by sramakintel** [comment]
> This PR will need some updates after the decision is made on how to add the scikit-learn-intelex package to it.

@elesel :Any updates on this PR ?

**Comment by rtanase** [comment]
> > This PR will need some updates after the decision is made on how to add the scikit-learn-intelex package to it.
> 
> @elesel :Any updates on this PR ?

@sramakintel, what was the reasoning behind creating a new container where intel extensions were added? What's downside of including in the existing one?

**Comment by sramakintel** [comment]
> > > This PR will need some updates after the decision is made on how to add the scikit-learn-intelex package to it.
> > 
> > 
> > @elesel :Any updates on this PR ?
> 
> @sramakintel, what was the reasoning behind creating a new container where intel extensions were added? What's downside of including in the existing one?

@rtanase: add a separate intel-extension folder to provide options for both intel-extension and stock version of sklearn based on user's preference. would you like to have them combined in a single Dockerfile and conda .yaml file?

**Comment by rtanase** [comment]
> > > > This PR will need some updates after the decision is made on how to add the scikit-learn-intelex package to it.
> > > 
> > > 
> > > @elesel :Any updates on this PR ?
> > 
> > 
> > @sramakintel, what was the reasoning behind creating a new container where intel extensions were added? What's downside of including in the existing one?
> 
> @rtanase: I added a separate intel-extension folder to provide options for both intel-extension and stock version of sklearn based on user's preference. would you like to have them combined in a single Dockerfile and conda .yaml file?

but unless the intel-extension is not imported or used at runtime, just installing it wouldn't affect in any way the stock version of sklearn, right? or would it?

**Comment by sramakintel** [comment]
> > > > > This PR will need some updates after the decision is made on how to add the scikit-learn-intelex package to it.
> > > > 
> > > > 
> > > > @elesel :Any updates on this PR ?
> > > 
> > > 
> > > @sramakintel, what was the reasoning behind creating a new container where intel extensions were added? What's downside of including in the existing one?
> > 
> > 
> > @rtanase: I added a separate intel-extension folder to provide options for both intel-extension and stock version of sklearn based on user's preference. would you like to have them combined in a single Dockerfile and conda .yaml file?
> 
> but unless the intel-extension is not imported or used at runtime, just installing it wouldn't affect in any way the stock version of sklearn, right? or would it?

correct. just installing it, wouldn't affect stock sklearn. Infact, intel-extension is disabled by default. More information here https://github.com/Azure/azureml-assets/blob/95855168c3e4a873df52a53674c4d63cc03a500a/training/general/src/environments/sklearn-1.1/context/README.md#run-a-linear-regression-workload-inside-the-container

File: training/general/src/environments/sklearn-1.1/context/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,5 @@
 FROM mcr.microsoft.com/azureml/openmpi4.1.0-ubuntu20.04:{{latest-image-tag}}
+# FROM mcr.microsoft.com/azureml/openmpi4.1.0-ubuntu20.04:latest
```
Comment (elesel) [inline training/general/src/environments/sklearn-1.1/context/Dockerfile:2]: Please remove this line.

File: training/general/src/environments/sklearn-1.1/context/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# SKLearn Docker Environment
```
Comment (elesel) [inline training/general/src/environments/sklearn-1.1/context/README.md:1]: Please move this file up one directory level to training/general/src/environments/sklearn-1.1.

File: training/general/src/environments/sklearn-1.1/context/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# SKLearn Docker Environment 
+
+## Overview
+This document provides details on how to create and run a sample workload in a containerized SKLearn Environment using stock version of SKLearn as well as Intel Extension for SKLearn on Azure ML instances.
+
+## How it works
+The container has stock and intel-optimized versions available. 
+To build the docker container use the `Dockerfile` and `conda_dependencies.yaml`.
+Refer to the link [here](https://github.com/Azure/azureml-assets/wiki/Environments#testing-environment-image-builds) on how the macros are replaced.
```
Comment (elesel) [inline training/general/src/environments/sklearn-1.1/context/README.md:9]: I don't think this is accurate anymore, now that the extension is being installed as part of the regular image. You could include information here on how to build the image locally but it would also help to describe how to use this environment in Azure ML.

File: training/general/src/environments/sklearn-1.1/tests/src/main.py [hunk 5]
```diff
@@ -1,8 +1,7 @@
 # Copyright (c) Microsoft Corporation.
```
Comment (elesel) [inline training/general/src/environments/sklearn-1.1/tests/src/main.py:1]: @rtanase @savitamittal1 Would you please review the changes to this script? Also, if we should test the operation of the Intel extension then a separate test should be created under the tests directory, either in sklearn_sample_test.py or another script.

File: training/general/src/environments/sklearn-1.1/tests/src/main.py [hunk 6]
```diff
@@ -1,8 +1,7 @@
...
 # Licensed under the MIT License.
-"""Simple Sklearn Test."""
```
Comment (elesel) [inline training/general/src/environments/sklearn-1.1/tests/src/main.py:3]: Please restore all of the removed comments.

File: training/general/src/environments/sklearn-1.1/tests/src/main.py [hunk 7]
```diff
@@ -1,8 +1,7 @@
...
+
 # imports
-import mlflow
```
Comment (elesel) [inline training/general/src/environments/sklearn-1.1/tests/src/main.py:5]: Why was mlflow removed?

File: training/general/src/environments/sklearn-1.1/tests/src/main.py [hunk 8]
```diff
@@ -13,17 +12,12 @@
...
 # define functions
 def main(args):
-    """Run and evaluate model."""
-    # enable auto logging
-    mlflow.autolog()
 
     # setup parameters
     params = {
         "fit_intercept": args.fit_intercept,
-        "normalize": args.normalize,
```
Comment (elesel) [inline training/general/src/environments/sklearn-1.1/tests/src/main.py:23]: Why was the normalize param removed?

**Review by elesel — CHANGES_REQUESTED** [review]

File: training/general/src/environments/sklearn-1.1/tests/src/main.py (hunk 8 above)
Comment (sramakintel) [reply training/general/src/environments/sklearn-1.1/tests/src/main.py:23]: Per latest docs of sklearn, https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html, `normalize` is no more a parameter.

File: training/general/src/environments/sklearn-1.1/context/Dockerfile (hunk 2 above)
Comment (sramakintel) [reply training/general/src/environments/sklearn-1.1/context/Dockerfile:2]: removed

File: training/general/src/environments/sklearn-1.1/context/README.md (hunk 3 above)
Comment (sramakintel) [reply training/general/src/environments/sklearn-1.1/context/README.md:1]: moved a directory up.

File: training/general/src/environments/sklearn-1.1/tests/src/main.py (hunk 6 above)
Comment (sramakintel) [reply training/general/src/environments/sklearn-1.1/tests/src/main.py:3]: restored

File: training/general/src/environments/sklearn-1.1/tests/src/main.py (hunk 7 above)
Comment (sramakintel) [reply training/general/src/environments/sklearn-1.1/tests/src/main.py:5]: added mlflow back

File: training/general/src/environments/sklearn-1.1/context/README.md (hunk 4 above)
Comment (sramakintel) [reply training/general/src/environments/sklearn-1.1/context/README.md:9]: @elesel : to build the docker image locally, we would have to replace macros in the dockerfile and conda yml file using command like `sed` which I had posted in an older version of the README.md which @rtanase had me remove it and pointed to the macros. Otherwise build will fail. 

Regarding, running on Azure ML, it is mentioned in the README, that the user would have to build the container and run `sklearn_sample_test.py` inside the container by logging in to an Azure ML instance. Do we want to add more details?

File: training/general/src/environments/sklearn-1.1/tests/src/main.py [hunk 9]
```diff
@@ -13,17 +12,12 @@
 
 # define functions
 def main(args):
-    """Run and evaluate model."""
-    # enable auto logging
-    mlflow.autolog()
```
Comment (rtanase) [inline training/general/src/environments/sklearn-1.1/tests/src/main.py:18]: was the mlflow.autolog() removed on purpose? why?

Comment (sramakintel) [reply training/general/src/environments/sklearn-1.1/tests/src/main.py:18]: @rtanase: we removed mlflow for some testing we were doing with sklearnex in the container, you can keep the line and mlflow related calls in the code - no concern. I reverted the changes to have mlflow included just as before

**Comment by sramakintel** [comment]
@elesel : any updates on status of this PR?

File: training/general/src/environments/sklearn-1.1/tests/src/main.py [hunk 10]
```diff
@@ -35,7 +31,6 @@ def main(args):
 
     print(model)
 
-
```
Comment (elesel) [inline training/general/src/environments/sklearn-1.1/tests/src/main.py:38]: Please restore this blank line to pass flake8 checks.

File: training/general/src/environments/sklearn-1.1/tests/src/main.py [hunk 11]
```diff
@@ -13,17 +12,14 @@
 
 # define functions
 def main(args):
-    """Run and evaluate model."""
+    """Run and evaluate model"""
```
Comment (elesel) [inline training/general/src/environments/sklearn-1.1/tests/src/main.py:15]: Reverting period removal to pass documentation checks.
```suggestion
    """Run and evaluate model."""
```

**Review by elesel — CHANGES_REQUESTED** [review]

File: training/general/src/environments/sklearn-1.1/tests/src/main.py (hunk 11 above)
Comment (sramakintel) [reply training/general/src/environments/sklearn-1.1/tests/src/main.py]: made changes

File: training/general/src/environments/sklearn-1.1/tests/src/main.py (hunk 10 above)
Comment (sramakintel) [reply training/general/src/environments/sklearn-1.1/tests/src/main.py]: changes made

File: training/general/src/environments/sklearn-1.1/context/README.md (hunk 4 above)
Comment (elesel) [reply training/general/src/environments/sklearn-1.1/context/README.md]: This works. Thanks!

**Review by elesel — APPROVED** [review]
Looks good to me. @rtanase, what do you think?

**Review by rtanase — APPROVED** [review]

**Comment by elesel** [comment]
@sramakintel Thanks for your contribution!
